File: android/core/src/main/kotlin/mesh/core/protocol/Codec.kt
/**
 * Mesh Messenger Core — Packet Codec
 *
 * Binary encode/decode for MeshPacket matching the wire format exactly.
 * Must produce byte-identical output to the TypeScript and Swift implementations
 * for interoperability (validated via golden test vectors).
 */

package mesh.core.protocol

import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Encode a MeshPacket into a binary frame.
 *
 * Layout:
 *   [0]       version (u8)
 *   [1]       type (u8)
 *   [2..9]    packetId (8 bytes)
 *   [10..25]  src NodeId (16 bytes)
 *   [26..41]  dst NodeId (16 bytes)
 *   [42]      ttl (u8)
 *   [43]      hopCount (u8)
 *   [44..47]  timestamp (u32 big-endian)
 *   [48..49]  payloadLength (u16 big-endian)
 *   [50..N]   payload
 *   [N+1..]   signature (optional)
 */
fun encodePacket(packet: MeshPacket): ByteArray {
    val payloadLen = packet.payload.size
    val sigLen = packet.signature?.size ?: 0
    val totalLen = MeshPacket.HEADER_SIZE + payloadLen + sigLen

    val buf = ByteBuffer.allocate(totalLen).order(ByteOrder.BIG_ENDIAN)

    buf.put(packet.version.toByte())
    buf.put(packet.type.value.toByte())
    buf.put(packet.packetId.bytes)
    buf.put(packet.src.bytes)
    buf.put(packet.dst.bytes)
    buf.put(packet.ttl.toByte())
    buf.put(packet.hopCount.toByte())
    buf.putInt(packet.timestamp.toInt())
    buf.putShort(payloadLen.toShort())
    buf.put(packet.payload)

    packet.signature?.let { buf.put(it) }

    return buf.array()
}

/**
 * Decode a binary frame into a MeshPacket.
 * @throws IllegalArgumentException on malformed input
 */
fun decodePacket(data: ByteArray): MeshPacket {
    require(data.size >= MeshPacket.HEADER_SIZE) {
        "Packet too short: ${data.size} < ${MeshPacket.HEADER_SIZE}"
    }

    val buf = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)

    val version = buf.get().toUByte()
    val type = PacketType.fromValue(buf.get().toInt())

    val packetIdBytes = ByteArray(8)
    buf.get(packetIdBytes)
    val packetId = PacketId(packetIdBytes)

    val srcBytes = ByteArray(16)
    buf.get(srcBytes)
    val src = NodeId(srcBytes)

    val dstBytes = ByteArray(16)
    buf.get(dstBytes)
    val dst = NodeId(dstBytes)

    val ttl = buf.get().toUByte()
    val hopCount = buf.get().toUByte()
    val timestamp = buf.getInt().toUInt()
    val payloadLength = buf.getShort().toInt() and 0xFFFF

    require(data.size >= MeshPacket.HEADER_SIZE + payloadLength) {
        "Payload truncated: expected $payloadLength bytes, got ${data.size - MeshPacket.HEADER_SIZE}"
    }

    val payload = ByteArray(payloadLength)
    buf.get(payload)

    val signature = if (buf.remaining() > 0) {
        ByteArray(buf.remaining()).also { buf.get(it) }
    } else null

    return MeshPacket(
        version = version,
        type = type,
        packetId = packetId,
        src = src,
        dst = dst,
        ttl = ttl,
        hopCount = hopCount,
        timestamp = timestamp,
        payload = payload,
        signature = signature,
    )
}

/**
 * Constant-time byte array comparison for MAC/fingerprint safety.
 */
fun constantTimeEquals(a: ByteArray, b: ByteArray): Boolean {
    if (a.size != b.size) return false
    var diff = 0
    for (i in a.indices) {
        diff = diff or (a[i].toInt() xor b[i].toInt())
    }
    return diff == 0
}

File: ios/Sources/MeshCore/Codec.swift
import Foundation

public enum CodecError: Error {
    case packetTooShort
    case payloadTruncated
    case invalidPacketType
}

/**
 * Encode a MeshPacket into a binary frame (Big-Endian).
 */
public func encodePacket(_ packet: MeshPacket) -> Data {
    var data = Data()
    
    data.append(packet.version)
    data.append(packet.type.rawValue)
    
    // Packet ID (8 bytes)
    data.append(packet.packetId.prefix(8))
    
    // Source & Dest NodeIDs (16 bytes each)
    data.append(packet.src.prefix(16))
    data.append(packet.dst.prefix(16))
    
    data.append(packet.ttl)
    data.append(packet.hopCount)
    
    // Timestamp (UInt32 big-endian)
    var timestampBE = packet.timestamp.bigEndian
    withUnsafeBytes(of: &timestampBE) { data.append(contentsOf: $0) }
    
    // Payload length (UInt16 big-endian)
    var lengthBE = UInt16(packet.payload.count).bigEndian
    withUnsafeBytes(of: &lengthBE) { data.append(contentsOf: $0) }
    
    // Payload
    data.append(packet.payload)
    
    // Signature (if present)
    if let sig = packet.signature {
        data.append(sig)
    }
    
    return data
}

/**
 * Decode a binary frame into a MeshPacket.
 */
public func decodePacket(_ data: Data) throws -> MeshPacket {
    guard data.count >= MeshPacket.headerSize else {
        throw CodecError.packetTooShort
    }
    
    let version = data[0]
    
    guard let type = PacketType(rawValue: data[1]) else {
        throw CodecError.invalidPacketType
    }
    
    let packetId = data.subdata(in: 2..<10)
    let src = data.subdata(in: 10..<26)
    let dst = data.subdata(in: 26..<42)
    
    let ttl = data[42]
    let hopCount = data[43]
    
    // Read timestamp
    let timestampBE: UInt32 = data.subdata(in: 44..<48).withUnsafeBytes {
        $0.load(as: UInt32.self)
    }
    let timestamp = UInt32(bigEndian: timestampBE)
    
    // Read payload length
    let lengthBE: UInt16 = data.subdata(in: 48..<50).withUnsafeBytes {
        $0.load(as: UInt16.self)
    }
    let payloadLength = Int(UInt16(bigEndian: lengthBE))
    
    guard data.count >= MeshPacket.headerSize + payloadLength else {
        throw CodecError.payloadTruncated
    }
    
    let payload = data.subdata(in: MeshPacket.headerSize..<(MeshPacket.headerSize + payloadLength))
    
    var signature: Data? = nil
    let signatureStart = MeshPacket.headerSize + payloadLength
    if data.count > signatureStart {
        signature = data.subdata(in: signatureStart..<data.count)
    }
    
    return MeshPacket(
        version: version,
        type: type,
        packetId: packetId,
        src: src,
        dst: dst,
        ttl: ttl,
        hopCount: hopCount,
        timestamp: timestamp,
        payload: payload,
        signature: signature
    )
}

/**
 * Constant-time data comparison for security.
 */
public func constantTimeEquals(_ a: Data, _ b: Data) -> Bool {
    guard a.count == b.count else { return false }
    var diff: UInt8 = 0
    for i in 0..<a.count {
        diff |= a[i] ^ b[i]
    }
    return diff == 0
}
